### PR TITLE
Added inter-task communication channels

### PR DESCRIPTION
**Pull request by amitmurthy**
This PR implements only the inter-task (not inter-process) channels part of https://github.com/JuliaLang/julia/pull/12042

The default channel (of size 32) is faster than produce-consume as seen below.   

``````
julia> function foo(t::Task)
           while true
               consume(t)
           end
       end;

julia> function foo(n::Int)
           self = current_task()
           @schedule foo(self)
           for i in 1:10^n
               produce(i)
           end
       end;

julia> foo(1);

julia> @time foo(6);
   2.226 seconds      (2005 k allocations: 31484 KB, 0.17% gc time)

julia> function bar(c::Channel)
           while true
               take!(c)
           end
       end;

julia> function bar(n, c)
           @schedule bar(c)
           for i in 1:10^n
               put!(c, i)
           end
       end;

julia> bar(1, Channel())

julia> @time bar(6, Channel())
 104.415 milliseconds (1125 k allocations: 21483 KB, 2.48% gc time)                                                                                                                      

julia> bar(1, Channel(1))

julia> @time bar(6, Channel(1))
   2.287 seconds      (5000 k allocations: 198 MB, 0.74% gc time)                                                                                                                        ```
``````

### CODE REVIEW

**Comment by amitmurthy** [comment]
Any objections to merging this? Both the feature and code is pretty straightforward and doesn't break any existing stuff.

I'll rebase #12042 if/when this gets merged. That will make is easier to review  #12042 and we can take a call if we want to go head with the relatively major change of #12042.

File: base/channels.jl [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    cond_put::Condition     # waiting for a writeable slot
+    state::Symbol
+
+    data::Array{T,1}
+    szp1::Int               # current channel size plus one
+    sz_max::Int             # maximum size of channel
+    take_pos::Int           # read position
+    put_pos::Int            # write position
+
+    function Channel(sz)
```
Comment (jakebolewski) [inline base/channels.jl]: I think this should throw an error rather the silently truncate.  I feel the user should also be able. To define a Chanel length greater than 32.

  I guess you could also throw a nicer error checking if the Chanel size is greater than 1.

File: doc/manual/parallel-computing.rst [hunk 2]
```diff
@@ -444,6 +444,22 @@ preemptively. This means context switches only occur at well-defined
...
+--------
+Channels provide for a fast means of inter-task communication. A
+``Channel(T::Type, n::Int)`` is a shared queue of maximum length ``n``
+holding objects of type ``T``. Multiple readers can read off the channel
+via ``fetch`` and ``take!``. Multiple writers can add to the channel via
+``put!``. ``isready`` tests for the prescence of any object in
+the channel, while ``wait`` waits for an object to become available.
+
+A Channel can be used as an iterable object in a ``for`` loop, in which
+case the loop runs till the channel is open. The loop variable takes on
```
Comment (IainNZ) [inline doc/manual/parallel-computing.rst]: Does "open" mean empty (queue length 0) here?

File: base/channels.jl (hunk 1 above)
Comment (amitmurthy) [inline base/channels.jl]: You can define a channel size greater than 32. `data` is a circular buffer that is managed via `szp1`, `take_pos` and `put_pos`. The initial allocation is 32, It is grown on demand to a maximum size of `sz_max`. Even a channel of `typemax(Int)` will only allocate storage for 32 objects initially.

File: doc/manual/parallel-computing.rst (hunk 2 above)
Comment (amitmurthy) [inline doc/manual/parallel-computing.rst]: It means empty queue and that `close` has not been called. The `for` loop will wait for more data even on a currently empty queue, as long as `close` has not been called. Similar to the `Task` iterator which runs till the task terminates.

File: base/channels.jl (hunk 1 above)
Comment (amitmurthy) [inline base/channels.jl]: A further optimization that should be done is halving `data` once the current number of objects is less than 50% of `data` length

File: doc/manual/parallel-computing.rst (hunk 2 above)
Comment (IainNZ) [inline doc/manual/parallel-computing.rst]: Ah I see. I think it needs to be `while` instead of `till` then.

Comment (amitmurthy) [inline doc/manual/parallel-computing.rst]: Noted.

**Comment by amitmurthy** [comment]
I'll merge this in a day or two if there are no further issues.

**Pull request merged**